packages/babel-core/src/config/files/module-types.ts

```diff
@@ -23,38 +27,82 @@ export const supportsESM = semver.satisfies(
   "^12.17 || >=13.2",
 );
 
-export default function* loadCjsOrMjsDefault(
+export default function* loadCodeDefault(
   filepath: string,
   asyncError: string,
   // TODO(Babel 8): Remove this
   fallbackToTranspiledModule: boolean = false,
 ): Handler<unknown> {
-  switch (guessJSModuleType(filepath)) {
-    case "cjs":
+  switch (path.extname(filepath)) {
+    case ".cjs":
       return loadCjsDefault(filepath, fallbackToTranspiledModule);
-    case "unknown":
+    case ".mjs":
+      break;
+    case ".cts":
+      return loadCtsDefault(filepath);
+    default:
       try {
         return loadCjsDefault(filepath, fallbackToTranspiledModule);
       } catch (e) {
         if (e.code !== "ERR_REQUIRE_ESM") throw e;
       }
-    // fall through
-    case "mjs":
-      if (yield* isAsync()) {
-        return yield* waitFor(loadMjsDefault(filepath));
-      }
-      throw new ConfigError(asyncError, filepath);
   }
+  if (yield* isAsync()) {
+    return yield* waitFor(loadMjsDefault(filepath));
+  }
+  throw new ConfigError(asyncError, filepath);
 }
 
-function guessJSModuleType(filename: string): "cjs" | "mjs" | "unknown" {
-  switch (path.extname(filename)) {
-    case ".cjs":
-      return "cjs";
-    case ".mjs":
-      return "mjs";
-    default:
-      return "unknown";
+function loadCtsDefault(filepath: string) {
+  const ext = ".cts";
+  const hasTsSupport = !!(
+    require.extensions[".ts"] ||
+    require.extensions[".cts"] ||
+    require.extensions[".mts"]
+  );
+  if (!hasTsSupport) {
+    const code = fs.readFileSync(filepath, "utf8");
+    const opts: InputOptions = {
+      babelrc: false,
+      configFile: false,
+      filename: path.basename(filepath),
+      sourceType: "script",
+      sourceMaps: "inline",
+      presets: [
+        [
+          "@babel/preset-typescript",
+          process.env.BABEL_8_BREAKING
+            ? {
+                disallowAmbiguousJSXLike: true,
+                allExtensions: true,
+                onlyRemoveTypeImports: true,
+                optimizeConstEnums: true,
+              }
+            : {
+                allowDeclareFields: true,
+                disallowAmbiguousJSXLike: true,
+                allExtensions: true,
+                onlyRemoveTypeImports: true,
+                optimizeConstEnums: true,
+              },
+        ],
+      ],
+    };
+    const result = transformSync(code, opts);
+    require.extensions[ext] = function (m, filename) {
+      if (filename === filepath) {
+        // @ts-expect-error Undocumented API
+        return m._compile(result.code, filename);
+      }
+      return require.extensions[".js"](m, filename);
+    };
+  }
+  try {
+    return endHiddenCallStack(require)(filepath);
+  } finally {
+    if (!hasTsSupport) {
```

Comment:
The user config could register a new handler for TS extensions. We should also check if `require.extensions[ext]` is still `===` to the function we defined a few lines above, before deleting it.